source/common/upstream/cluster_manager_impl.h

```diff
@@ -260,6 +263,7 @@ class ClusterManagerImpl : public ClusterManager, Logger::Loggable<Logger::Id::u
     Event::Dispatcher& thread_local_dispatcher_;
     std::unordered_map<std::string, ClusterEntryPtr> thread_local_clusters_;
     std::unordered_map<HostConstSharedPtr, ConnPoolsContainer> host_http_conn_pool_map_;
+    std::set<Envoy::Upstream::ClusterUpdateCallbacks*> update_callbacks_;
```

Comment:
unordered_set?

Reply:
That is how I had it initially but I was decided that an ordered set would make debugging easier.

Reply:
Here is what I would personally do here which will solve this comment as well as my other comment about RAII:

1) Use an `std::list` which gives O(1) add. Coupled with https://github.com/envoyproxy/envoy/blob/master/source/common/common/linked_object.h you can also get O(1) removal.
2) Instead of having an explicit "remove" API, return an opaque handle, which itself just contains the linked entry.
3) When the opaque handle is destroyed, the callback is unlinked. 

I think this is the optimal solution and pretty easy to implement. WDYT?

Reply:
Done.